lib/publiccloud/utils.pm

```diff
@@ -859,4 +860,70 @@ sub detect_worker_ip {
     die "Worker IP could not be determined - return was $ip";
 }
 
+
+
+=head2 trup_call_remote 
+
+    trup_call_remote(command, [timeout, exit_code, proceed_on_failure])
+
+    Run a transactional command on remote machine via ssh, non-interactive
+
+    Return: 
+    ($out, $res): a string with the full screen output and the number of final exit code: 
+
+=cut
+
+sub trup_call_remote {
+    my ($self, $cmd, %args) = @_;
+    my $start = time();
+    my $script = "sudo transactional-update -n $cmd";
+    my $timeout = $args{timeout} // 600;
+    my $exit = $args{exit_code} // 0;
+    my $proceed = $args{proceed_on_failure} // 0;
+    my ($out, $res);
+    # Run command, add exit code at end
+    $out = $self->ssh_script_output($script . "; echo exit-$?-", timeout => $args{timeout}, proceed_on_failure => $proceed);
+    $res = get_transact_up_result($out, $cmd);
+    record_info("ssh tr-up", "ssh tr-up $cmd ($res), duration " . (time() - $start) . "sec.\n", result => ($res == $exit) ? 'ok' : 'fail');
+    return ($out, $res) if ($res == 0 || $proceed);
+    die "transactional-update didn't finish" unless defined($res);
+    die "transactional-update returned with $res, expected $exit" unless ($res == $exit);
+}
+
+=head2 get_transact_up_result 
+
+    get_transact_up_result($screen): 
+        transactional-update may exit ok with 0 even with errors in the process:
+        so, here we parse its output, to provide a contextual final result.
+
+    Check in the screen output: no-final error, 'finished' message present, no middle error, only 'exit code 0'.
+    Handle final result string 'exit-$?-', if eventually added to the output.
+    No exit code is considered ok, like in nothing to update.
+
+    Return: 1 = fail; 0 = pass; empty = no result.
+
+=cut
+
+sub get_transact_up_result {
+    my ($screen, $cmd) = @_;
+    return unless ($screen);
+    my @lines = split(/\n/, $screen);
+    $lines[-1] =~ /^exit-(\d+)-$/ // 0;
+    my $exit = $1 // -1;
+    # error in final result
+    return 1 if ($exit > 0);
+    return 1 if ($cmd =~ /^reboot/ && $screen !~ /Requesting reboot using/);
```

Comment:
Where are we calling reboot?

Reply:
> NO, please. I checked and tested it and the code is working as it is

Then let's add Perl unit tests for this in the t/ directory.  I'm not comfortable adding library code that's not tested.

Reply:
routine removed